### PR TITLE
Introducing IHtmlContent in Mvc.Razor.

### PR DESCRIPTION
**Pull request by sornaks**
- Changing HtmlHelper and Helper result to implement IHtmlContent.
- Introducing BufferedHtmlContent.
- Making RazorPage handle only IHtmlContent and clearing out other types.
- Making StringCollectionTextWriter use BufferedHtmlContent so that it can be returned where necessary.
- Updating places which involve Write/Copy to pass in encoders.
- The encoders are currently not being used during write. But when HtmlString is modified to carry encode metadata, the encoder can be used for writing. This is a perf optimization and hence not a part of this change.

### CODE REVIEW

**Comment by sornaks** [comment]
Haven't added changes which include updating `TagHelperContent` to implement `IHtmlContent`. I'll do that tomorrow. /cc: @rynowak

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/BufferedHtmlContent.cs [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        }
+
+        /// <summary>
+        /// Appends a character array to the collection.
+        /// </summary>
+        /// <param name="value">The character array to be appended.</param>
+        /// <param name="index">The index from which the character array must be read.</param>
+        /// <param name="count">The count till which the character array must be read.</param>
+        /// <remarks>
+        /// Splits the character array into strings of 1KB length and appends them.
```
Comment (IDisposable) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/BufferedHtmlContent.cs]: Should we be specifying 1KB here when it's actually MaxCharToStringLength. Seems like an implementation detail that is being leaked.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/BufferedHtmlContent.cs [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+                    writer.Write(entryAsString);
+                }
+                else
+                {
+                    // Only string, IHtmlContent values can be added to the buffer.
+                    var entryAsIHtmlContent = (IHtmlContent)entry;
+                    if (entryAsIHtmlContent != null)
+                    {
+                        entryAsIHtmlContent.WriteTo(writer, encoder);
+                    }
```
Comment (IDisposable) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/BufferedHtmlContent.cs]: else Debug.Assert?

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/BufferedHtmlContent.cs [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+                if (entryAsString != null)
+                {
+                    // encoder.HtmlEncode(entryAsString, writer);
+                    writer.Write(entryAsString);
+                }
+                else
+                {
+                    // Only string, IHtmlContent values can be added to the buffer.
+                    var entryAsIHtmlContent = (IHtmlContent)entry;
+                    if (entryAsIHtmlContent != null)
```
Comment (IDisposable) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/BufferedHtmlContent.cs]: We could throw here with for an entry that is a ```string``` but is null. Since we're exposing the ```List<object>``` as an internal property we don't know going forward that nobody is going to do that... perhaps the whole thing should be based on ```is string``` and ```is IHtmlContent``` rather that throwing an ```InvalidCastException```

File: src/Microsoft.AspNet.Mvc.Razor/RazorTextWriter.cs [hunk 4]
```diff
@@ -51,6 +52,8 @@ public override Encoding Encoding
 
         private TextWriter TargetWriter { get; set; }
 
+        private IHtmlEncoder HtmlEncoder { get; set; } = new HtmlEncoder();
```
Comment (IDisposable) [inline src/Microsoft.AspNet.Mvc.Razor/RazorTextWriter.cs]: Are ```HtmlEncoders``` both cheap to new and stateful? If not **both**, perhaps we should cache a static Instance for defaulting? Even if they are stateful, but still expensive to new, perhaps a lazy creation?

File: src/Microsoft.AspNet.Mvc.Razor/RazorView.cs [hunk 5]
```diff
@@ -205,7 +206,7 @@ private bool EnableInstrumentation
             if (bodyWriter.IsBuffering)
             {
                 // Only copy buffered content to the Output if we're currently buffering.
-                await bodyWriter.CopyToAsync(context.Writer);
+                await bodyWriter.CopyToAsync(context.Writer, new HtmlEncoder());
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Mvc.Razor/RazorView.cs]: Why are we newing this up everywhere? Don't we get it from DI?

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/BufferedHtmlContent.cs (hunk 3 above)
Comment (sornaks) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/BufferedHtmlContent.cs]: We block 'null' 'string's and 'IHtmlContent' in 'Append'.

File: src/Microsoft.AspNet.Mvc.Razor/HelperResult.cs [hunk 6]
```diff
@@ -3,30 +3,33 @@
...
 namespace Microsoft.AspNet.Mvc.Razor
 {
     /// <summary>
     /// Represents a deferred write operation in a <see cref="RazorPage"/>.
     /// </summary>
-    public class HelperResult
+    public class HelperResult : IHtmlContent
     {
-        private readonly Action<TextWriter> _action;
+        private readonly Action<TextWriter, IHtmlEncoder> _action;
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor/HelperResult.cs]: Would it be simpler to just leave this as `Func<TextWriter>`? We're planning to delete this class anyway. How big of a change would that be to include in this PR?

File: src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs [hunk 7]
```diff
@@ -449,7 +445,7 @@ public virtual void WriteTo([NotNull] TextWriter writer, object value)
                     //
                     // Do not combine following condition with check of escapeQuotes; htmlString.ToString() can be
                     // expensive when the HtmlString is created with a StringCollectionTextWriter.
-                    var stringValue = htmlString.ToString();
+                    var stringValue = htmlContent.ToString();
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: I'm not happy to see a `ToString()` - but I get why it's showing up. Let's just make a mental note to revist as part of the perf part of this change.

File: src/Microsoft.AspNet.Mvc.Razor/RazorTextWriter.cs (hunk 4 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor/RazorTextWriter.cs]: They are pretty cheap, but this is wrong for a different reason - this is a service that we get from DI.

File: src/Microsoft.AspNet.Mvc.Razor/RazorTextWriter.cs [hunk 8]
```diff
@@ -186,24 +189,24 @@ public override void Flush()
...
         }
 
         /// <inheritdoc />
-        public Task CopyToAsync(TextWriter writer)
+        public Task CopyToAsync(TextWriter writer, IHtmlEncoder encoder)
         {
             writer = UnWrapRazorTextWriter(writer);
-            return BufferedWriter.CopyToAsync(writer);
+            return BufferedWriter.CopyToAsync(writer, encoder);
         }
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor/RazorTextWriter.cs]: Can you revert the changes to this class? What breaks? No-one should be calling `TextWriter.Write(...)` and passing in an `IHtmlContent`

File: src/Microsoft.AspNet.Mvc.Razor/RazorView.cs (hunk 5 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor/RazorView.cs]: This should be available already somehow.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/BufferedHtmlContent.cs [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+        /// <inheritdoc />
+        public void WriteTo([NotNull] TextWriter writer, [NotNull] IHtmlEncoder encoder)
+        {
+            foreach (var entry in Entries)
+            {
+                var entryAsString = entry as string;
+                if (entryAsString != null)
+                {
+                    // encoder.HtmlEncode(entryAsString, writer);
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/BufferedHtmlContent.cs]: bleh

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/BufferedHtmlContent.cs (hunk 3 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/BufferedHtmlContent.cs]: So then don't have the null-check since it can't happen.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/BufferedHtmlContent.cs [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+using Microsoft.AspNet.Html.Abstractions;
+using Microsoft.Framework.Internal;
+using Microsoft.Framework.WebEncoders;
+
+namespace Microsoft.AspNet.Mvc.Rendering
+{
+    /// <summary>
+    /// Enumerable object collection which knows how to write itself.
+    /// </summary>
+    public class BufferedHtmlContent : IHtmlContent, IEnumerable<object>
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/BufferedHtmlContent.cs]: Do we need `IEnumerable<object>`? Who's looking at it?

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/StringCollectionTextWriter.cs [hunk 11]
```diff
@@ -28,7 +29,7 @@ public class StringCollectionTextWriter : TextWriter
         public StringCollectionTextWriter(Encoding encoding)
         {
             _encoding = encoding;
-            Buffer = new BufferEntryCollection();
+            Content = new BufferedHtmlContent();
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/StringCollectionTextWriter.cs]: Not sure about the need for changes to this class. It's a `TextWriter` and as such it can safely just operate on strings.

File: src/Microsoft.AspNet.PageExecutionInstrumentation.Interfaces/IBufferedTextWriter.cs [hunk 12]
```diff
@@ -19,14 +20,16 @@ public interface IBufferedTextWriter
...
 
         /// <summary>
         /// Asynchronously copies the buffered content to the <paramref name="writer"/>.
         /// </summary>
-        /// <param name="writer">The writer to copy the contents to.</param>
+        /// <param name="writer">The <see cref="TextWriter"/> to copy the contents to.</param>
+        /// <param name="encoder">The <see cref="IHtmlEncoder"/> which encodes the content to be written.</param>
         /// <returns>A <see cref="Task"/> representing the copy operation.</returns>
-        Task CopyToAsync(TextWriter writer);
+        Task CopyToAsync(TextWriter writer, IHtmlEncoder encoder);
```
Comment (rynowak) [inline src/Microsoft.AspNet.PageExecutionInstrumentation.Interfaces/IBufferedTextWriter.cs]: We shouldn't change this api

**Comment by rynowak** [comment]
:watch:

File: src/Microsoft.AspNet.PageExecutionInstrumentation.Interfaces/IBufferedTextWriter.cs (hunk 12 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.PageExecutionInstrumentation.Interfaces/IBufferedTextWriter.cs]: We'll need to let the BrowserLink folks know if any API changes are made in this project.

Comment (marcosph) [inline src/Microsoft.AspNet.PageExecutionInstrumentation.Interfaces/IBufferedTextWriter.cs]: I have an MVC6 intranet application that uses roles. The admin users can
manage all the user roles. However any changes made to a users' roles are
only effective when the user logs on (this makes sense as the roles are
stored as claims in a cookie). The requirement is for role changes to take
effect immediately, like they did in the older asp.net membership and role
managers. How can one achieve this with the new identity models? The only
way that comes to mind is to set some flag that is checked on each request
and if set then update the cookie to reflect the new roles. Do you have any
other suggestions or examples of how to ensure the role changes take effect
immediatley?

Comment (sornaks) [inline src/Microsoft.AspNet.PageExecutionInstrumentation.Interfaces/IBufferedTextWriter.cs]: @marcosph can you file an issue for this? We can discuss it there.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/StringCollectionTextWriter.cs (hunk 11 above)
Comment (sornaks) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/StringCollectionTextWriter.cs]: This is changing because we removed the `StringCollectionTextWriter` constructor in `HtmlString` which is used here - https://github.com/aspnet/Mvc/pull/2797/files#diff-0e5ef27109c7145789f12182064f84acL423

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/BufferedHtmlContent.cs (hunk 10 above)
Comment (sornaks) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/BufferedHtmlContent.cs]: We dont. Let me remove it.

File: src/Microsoft.AspNet.Mvc.Razor/RazorTextWriter.cs (hunk 8 above)
Comment (sornaks) [inline src/Microsoft.AspNet.Mvc.Razor/RazorTextWriter.cs]: This is the place it is used - https://github.com/aspnet/Mvc/pull/2797/files#diff-21281ad55491937a33b7eee9b50bdc81L278

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/BufferedHtmlContent.cs (hunk 9 above)
Comment (sornaks) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/BufferedHtmlContent.cs]: The comment was a mistake. But this code will be present with the perf change (when encoding is delayed).

File: src/Microsoft.AspNet.Mvc.Razor/HelperResult.cs (hunk 6 above)
Comment (sornaks) [inline src/Microsoft.AspNet.Mvc.Razor/HelperResult.cs]: It shouldn't be a big change. Let me try getting it done.

**Comment by sornaks** [comment]
Updated..

File: src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs [hunk 13]
```diff
@@ -395,7 +388,8 @@ public virtual void Write(object value)
         /// <param name="value">The <see cref="object"/> to write.</param>
         /// <remarks>
         /// <paramref name="value"/>s of type <see cref="HtmlString"/> are written without encoding and the
-        /// <see cref="HelperResult.WriteTo(TextWriter)"/> is invoked for <see cref="HelperResult"/> types.
+        /// <see cref="HelperResult.WriteTo(TextWriter, IHtmlEncoder)"/> is invoked for
+        /// <see cref="HelperResult"/> types.
         /// For all other types, the encoded result of <see cref="object.ToString"/> is written to the
         /// <paramref name="writer"/>.
         /// </remarks>
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: These docs are out of date, these should just refer to `IHtmlContent`

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/HtmlString.cs [hunk 14]
```diff
@@ -39,29 +37,26 @@ public static HtmlString Empty
...
+            var writerAsStringCollectionTextWriter = writer as StringCollectionTextWriter;
+            if (writerAsStringCollectionTextWriter != null)
             {
-                _writer.CopyTo(targetWriter);
+                writerAsStringCollectionTextWriter.Content.Append(_input);
             }
             else
             {
-                targetWriter.Write(_input);
+                writer.Write(_input);
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/HtmlString.cs]: What does this actually accomplish? I would expect `writerAsStringCollectionTextWriter.Content.Append(_input)` and `writer.Write(_input)` to do the same thing

File: src/Microsoft.AspNet.PageExecutionInstrumentation.Interfaces/project.json [hunk 15]
```diff
@@ -15,5 +15,8 @@
                 "System.IO": "4.0.10-beta-*"
             }
         }
+    },
+    "dependencies": {
+        "Microsoft.Framework.WebEncoders.Core": "1.0.0-*"
```
Comment (rynowak) [inline src/Microsoft.AspNet.PageExecutionInstrumentation.Interfaces/project.json]: Remove

File: test/Microsoft.AspNet.Mvc.Razor.Test/RazorTextWriterTest.cs [hunk 16]
```diff
@@ -228,9 +231,9 @@ public void Copy_CopiesContent_IfTargetTextWriterIsARazorTextWriterAndBuffering(
 
             // Assert
             // Make sure content was written to the source.
-            Assert.Equal(2, source.BufferedWriter.Buffer.BufferEntries.Count);
-            Assert.Equal(1, target.BufferedWriter.Buffer.BufferEntries.Count);
-            Assert.Same(source.BufferedWriter.Buffer.BufferEntries, target.BufferedWriter.Buffer.BufferEntries[0]);
+            Assert.Equal(2, source.BufferedWriter.Content.Entries.Count());
+            Assert.Equal(1, target.BufferedWriter.Content.Entries.Count());
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.Razor.Test/RazorTextWriterTest.cs]: Isn't this `IList`?

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/BufferedHtmlContentTest.cs [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+namespace Microsoft.AspNet.Mvc.Rendering
+{
+    public class BufferedHtmlContentTest
+    {
+        [Fact]
+        public void AppendString_AppendsAString()
+        {
+            // Arrange
+            var content = new BufferedHtmlContent();
+            var writer = new StringWriter();
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/BufferedHtmlContentTest.cs]: unused

Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/BufferedHtmlContentTest.cs]: Looks like there are more of these, clean it up!

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/BufferedHtmlContentTest.cs [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+
+            // Act
+            content.Append(new TestHtmlContent("Hello"));
+
+            // Assert
+            var result = Assert.Single(content.Entries);
+            var testHtmlContent = Assert.IsType<TestHtmlContent>(result);
+            testHtmlContent.WriteTo(writer, new CommonTestEncoder());
+            Assert.Equal("Written from TestHtmlContent: Hello", writer.ToString());
+            writer.Dispose();
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/BufferedHtmlContentTest.cs]: No need to dispose a `StringWriter`

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/BufferedHtmlContentTest.cs [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+            var writer = new StringWriter();
+            content.Append(new TestHtmlContent("Hello"));
+            content.Append("Test");
+
+            // Act
+            content.WriteTo(writer, new CommonTestEncoder());
+
+            // Assert
+            Assert.Equal(2, content.Entries.Count());
+            Assert.Equal("Written from TestHtmlContent: HelloTest", writer.ToString());
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/BufferedHtmlContentTest.cs]: How is this different from `CanAppendMultipleItems`?

**Comment by rynowak** [comment]
Looking better. :watch:

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/HtmlString.cs (hunk 14 above)
Comment (sornaks) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/HtmlString.cs]: Yes you are right. It does the same thing.

File: test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/BufferedHtmlContentTest.cs (hunk 19 above)
Comment (sornaks) [inline test/Microsoft.AspNet.Mvc.ViewFeatures.Test/Rendering/BufferedHtmlContentTest.cs]: I'll make the other test not use a writer.. They are similar but with different purposes.

**Comment by sornaks** [comment]
Updated..

File: src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs [hunk 20]
```diff
@@ -449,7 +435,7 @@ public virtual void WriteTo([NotNull] TextWriter writer, object value)
                     //
                     // Do not combine following condition with check of escapeQuotes; htmlString.ToString() can be
                     // expensive when the HtmlString is created with a StringCollectionTextWriter.
```
Comment (kirthik) [inline src/Microsoft.AspNet.Mvc.Razor/RazorPage.cs]: comment should be changed. It mentions `htmlString.ToString()` which does not exist anymore. Also change `HtmlString` to `IHtmlContent`

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/BufferedHtmlContent.cs [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+using Microsoft.Framework.WebEncoders;
+
+namespace Microsoft.AspNet.Mvc.Rendering
+{
+    /// <summary>
+    /// Enumerable object collection which knows how to write itself.
+    /// </summary>
+    public class BufferedHtmlContent : IHtmlContent
+    {
+        private const int MaxCharToStringLength = 1024;
```
Comment (kirthik) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/BufferedHtmlContent.cs]: Add a new line.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/HtmlString.cs [hunk 22]
```diff
@@ -39,29 +37,18 @@ public static HtmlString Empty
...
         /// <summary>
         /// Writes the value in this instance of <see cref="HtmlString"/> to the target
-        /// <paramref name="targetWriter"/>.
+        /// <paramref name="writer"/>.
         /// </summary>
-        /// <param name="targetWriter">The <see cref="TextWriter"/> to write contents to.</param>
-        public void WriteTo(TextWriter targetWriter)
+        /// <param name="writer">The <see cref="TextWriter"/> to write contents to.</param>
+        /// <param name="encoder">The <see cref="IHtmlEncoder"/> with which the output must be encoded.</param>
+        public void WriteTo(TextWriter writer, IHtmlEncoder encoder)
```
Comment (kirthik) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/HtmlString.cs]: Is encoder needed here as a paramater? Looks like we are not using it in the method?

Comment (sornaks) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/HtmlString.cs]: Yes. It is coming from the interface `IHtmlContent`. We currently encode the text beforehand. Hence its not being used here. I have a change coming, which delays encoding until it is written.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/StringCollectionTextWriter.cs [hunk 23]
```diff
@@ -41,12 +42,12 @@ public override Encoding Encoding
...
         /// </summary>
         // internal for testing purposes.
-        internal BufferEntryCollection Buffer { get; }
+        internal BufferedHtmlContent Content { get; }
 
         /// <inheritdoc />
         public override void Write(char value)
         {
-            Buffer.Add(value.ToString());
+            Content.Append(value.ToString());
```
Comment (KrzysztofCwalina) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/StringCollectionTextWriter.cs]: Why do you call ToString here? BufferedHtmlContent has an overload to add a char.

Comment (sornaks) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/StringCollectionTextWriter.cs]: It is a character array which in turn is converted to a string. Instead of char -> char[] -> string it is just char -> string here.

Comment (KrzysztofCwalina) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/StringCollectionTextWriter.cs]: Ah, for some reason I thought you had a char overload. Anyway, did you consider storing more types (char[], char) "natively" in the Entries array so that you can write them to the textwriter without creating string first?

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/BufferedHtmlContent.cs [hunk 24]
```diff
@@ -0,0 +1,113 @@
...
+        // This is not List<IHtmlContent> because that would lead to boxing all strings to IHtmlContent
+        // which is not space performant.
+        // internal for testing.
+        internal List<object> Entries { get; set; } = new List<object>();
```
Comment (Eilon) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/BufferedHtmlContent.cs]: Can remove the `set;` right? I didn't see anything calling it.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/HtmlString.cs [hunk 25]
```diff
@@ -2,33 +2,31 @@
...
+        /// </summary>
+        /// <param name="input"><c>string</c>to initialize <see cref="HtmlString"/>.</param>
         public HtmlString(string input)
         {
             _input = input;
         }
 
         /// <summary>
-        /// Initializes a new instance of <see cref="HtmlString"/> that is backed by <paramref name="writer"/>.
+        /// Returns a <see cref="HtmlString"/> with empty content.
```
Comment (Eilon) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/HtmlString.cs]: `a` -> `an` :smile:

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/StringCollectionTextWriter.cs (hunk 23 above)
Comment (sornaks) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/StringCollectionTextWriter.cs]: Yes, the reason why `char` is not present is because it is not a widely used overload in razor (`TagBuilder` uses it for adding some consts. I'm changing this in an upcoming PR). And the reason why `char[]` is not present is to avoid calling `TextWriter.Write()`s on each character and instead have it call `Write()` on the entire string once.
In general I'd prefer keeping `BufferedHtmlContent` simple; otherwise we'll have to worry about handling cases for each type we support. The only reason why it has a `List<object>` and not a `List<IHtmlContent>` is because we have to box strings everytime we store them (and strings are widely used in Razor).

Comment (KrzysztofCwalina) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/StringCollectionTextWriter.cs]: But TextWriter has Write overloads that write char[] (not char by char)

**Comment by rynowak** [comment]
:shipit: from me. Take a note for @KrzysztofCwalina 's suggestion so we can follow up on it when we do the perf part of this change.

**Comment by sornaks** [comment]
Sure will do.

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/BufferedHtmlContent.cs [hunk 26]
```diff
@@ -0,0 +1,113 @@
...
+
+namespace Microsoft.AspNet.Mvc.Rendering
+{
+    /// <summary>
+    /// Enumerable object collection which knows how to write itself.
+    /// </summary>
+    public class BufferedHtmlContent : IHtmlContent
+    {
+        private const int MaxCharToStringLength = 1024;
+        // This is not List<IHtmlContent> because that would lead to boxing all strings to IHtmlContent
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/BufferedHtmlContent.cs]: `string` is not a value type. Do you mean "wrap"?

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/BufferedHtmlContent.cs (hunk 24 above)
Comment (sornaks) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/BufferedHtmlContent.cs]: Yes. Changed it.

**Comment by sornaks** [comment]
Checked in. Thanks! https://github.com/aspnet/Mvc/commit/a02082397a4e012b925a0f9ffff86ee54259812a

File: src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/BufferedHtmlContent.cs (hunk 26 above)
Comment (sornaks) [inline src/Microsoft.AspNet.Mvc.ViewFeatures/Rendering/BufferedHtmlContent.cs]: Did not see this comment before check in. Yes. Wrap. I'll update this in the next PR.
